Source/Details/ASCollectionLayoutCalculating.h

```diff
@@ -0,0 +1,64 @@
+//
+//  ASCollectionLayoutCalculating.h
+//  AsyncDisplayKit
+//
+//  Created by Huy Nguyen on 28/2/17.
+//  Copyright © 2017 Facebook. All rights reserved.
+//
+
+#import <Foundation/Foundation.h>
+#import <AsyncDisplayKit/ASDimension.h>
+
+@class ASLayout, ASElementMap;
+
+NS_ASSUME_NONNULL_BEGIN
+
+@protocol ASCollectionLayoutCalculating <NSObject>
+
+/**
+ * Whether or not size ranges must be provided via collection element in the element map.
+ * Default to NO, which means the calculator can determine size range for each collection element.
+ * ASLegacyCollectionLayoutCalculator might return YES because size ranges used to be provided via async delegate of table/collection view or layout inspector of collection view.
+ */
+@property (nonatomic, assign, readonly) BOOL sizeRangesShouldBeProvided;
+
+/**
+ * @abstract Return a layout of elements.
+ *
+ * @param viewportSize The viewport size that the resulting layout should fit, usually equals to the bounds of the containing table/collection view.
+ *
+ * @param map An element map containings all elements to be laid out.
+ *
+ * @discussion This method can be called on multiple theads and thus must be thread-safe (and ideally, stateless).
+ * This method must block the calling thread but can dispatch to other theads to reduce the blocking time.
+ */
+- (ASLayout *)layoutThatFits:(CGSize)viewportSize forElementMap:(ASElementMap *)map;
```

Comment:
(Sorry to spam) Should this object also synchronously decide what supplementary elements exist? Invalidation is a tough question here, but as a first stab:

```objc
- (void)supplementaryElementsForMap:(ASElementMap *)map storage:(id<ASSupplementaryElementStorage>)storage {
   // code
   [storage addSupplementaryElementOfKind:kind atIndexPath:indexPath];
   // code
}
```

We would call that method synchronously after the user issues each update, so that whatever supplementary elements they specify, we can go get the nodeBlocks for them as needed. Invalidating supplementary elements is a tough question and we should probably shelve it for now. 

If we add that, then we can fully-replace the layout inspector.